Fix #8192: Added search bar to classroom page

aks681 opened the pull request:
## Explanation

This PR fixes #8192 by adding a search bar to the bottom of the classroom page.

Screenshot on desktop:
![Annotation 2019-12-23 180254](https://user-images.githubusercontent.com/22347970/71358759-167d3e80-25b0-11ea-9ccd-ca0ea8b98ff5.jpg)

Mobile:
![image](https://user-images.githubusercontent.com/22347970/71358786-2ac13b80-25b0-11ea-8338-27c6f4b481c2.png)


## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

aks681 committed:
added search bar to classroom page (fix #8192)

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into classroom-page

aks681 committed:
added translations

kevinlee12 commented on core/templates/dev/head/pages/library-page/search-bar/search-bar.directive.ts:
```diff
@@ -32,7 +32,9 @@ angular.module('oppia').directive('searchBar', [
       scope: {},
-      bindToController: {},
+      bindToController: {
+        dropup: '&'
```
Is there a reason why it's named this way? (it sounds weird, but I'm okay if it's from convention).

aks681 replied:
Actually, it was for flipping the dropdown list in the search bar to a dropup list, as in the initial screenshot. I'll rename it to make it clearer.

seanlip commented on assets/i18n/en.json:
```diff
@@ -44,6 +44,8 @@
     "I18N_ACTION_CREATE_LESSON": "Create your own Lesson",
     "I18N_CARD_HEIGHT_LIMIT_WARNING_MESSAGE": "This card is quite long, and students might lose interest. Consider shortening it, or breaking it into two cards.",
     "I18N_CLASSROOM_PAGE_HEADING": "The Oppia Classroom Page",
+    "I18N_CLASSROOM_PAGE_SEARCH_BAR_HEADING": "Explore More Lessons:",
```
Let's drop the colon at the end, and make the "Made by the Community" text a little smaller.

aks681 replied:
Done

seanlip reviewed (changes requested):
Thanks, looks good! We'll probably need to do a bit of visual cleanup at some stage but I think this is good as a first step.

One question: could you please add some e2e tests for this flow? We should make sure that the overall user journey works correctly, now and in the future.

Also, could you explain a bit about what's going on with the "dropup" thing? If I'm reading correctly, its value doesn't seem to actually get used.

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into classroom-page

aks681 commented:
@seanlip The `dropup` value is used in the html file to change the dropdown list in the search bar to a dropup list so that the options are visible. I'll rename it to make it clearer.

aks681 committed:
added initial e2e test [testing]

aks681 committed:
made review changes

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into classroom-page

aks681 committed:
made subheading smaller

aks681 committed:
fixed e2e test

seanlip commented:
I understand the concept, but where is it used? It doesn't seem to be a valid HTML5 attribute.

aks681 committed:
changed year to 2020 :)

aks681 commented on core/templates/dev/head/pages/library-page/search-bar/search-bar.directive.html:
```diff
@@ -64,7 +64,7 @@
   </div>
 </div>
-<div ng-class="{'open' : $ctrl.activeMenuName === 'language'}" uib-dropdown class="float-left oppia-navbar-button-container oppia-search-bar-language-selector protractor-test-search-bar-language-selector">
+<div ng-class="{'open' : $ctrl.activeMenuName === 'language', 'dropup' : $ctrl.dropup()}" uib-dropdown class="float-left oppia-navbar-button-container oppia-search-bar-language-selector protractor-test-search-bar-language-selector">
```
@seanlip I use it here to enable the `dropup` class of `uib-dropdown`.

seanlip replied:
Ah, got it. Thanks! I did not realize it was a Bootstrap thing.

aks681 committed:
Completed e2e test

aks681 committed:
fix lint

seanlip commented:
Hi @aks681 -- just to check, this isn't ready for review yet, right? I noticed that one comment above still hasn't been replied to.

aks681 commented:
Addressed the comment. Modified screenshot:
![image](https://user-images.githubusercontent.com/22347970/71624349-13b1c780-2c08-11ea-81f4-2b322901d4ca.png)

It's ready for review now!

aks681 commented:
Also, I added the e2e test to circleci intead of travis, since I saw that the other e2e tests that use file upload and hence, does not use `prod_env`, are also in circleci.

seanlip commented on core/tests/protractor_desktop/classroomPage.js:
```diff
@@ -0,0 +1,102 @@
+        'The set of topic IDs for each classroom page.',
+        'List',
+        function(elem) {
+          elem.editItem(0, 'Dictionary').editEntry(1, 'List').addItem(
```
Just a question, out of curiosity -- why is the index for the list 1 rather than 0?

aks681 replied:
That is the index of the entry in the dictionary, the 0th of which is the classroom name and 1st, the list of topic IDs, and the initial 0 chooses that entry.

seanlip reviewed (approved):
LGTM, thanks!

seanlip commented:
@kevinlee12 @nithusha21 PTAL as codeowners, thanks!

kevinlee12 commented on core/tests/protractor_utils/ClassroomPage.js:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
```
Year :)

aks681 replied:
Done, maybe the first file with 2020 :).

kevinlee12 replied:
Looks like it, happy new year!

aks681 replied:
You too!

kevinlee12 reviewed (approved):
lgtm! (just a nit).

aks681 commented:
@nithusha21 PTAL as codeowner?

nithusha21 reviewed (approved):
LGTM thanks! Sorry for the delay in reviewing!
